vector-stores/spring-ai-couchbase-store/src/main/java/org/springframework/ai/vectorstore/CouchbaseVectorStore.java

```diff
@@ -0,0 +1,417 @@
+		this.bucket = cluster.bucket(config.bucketName);
+		this.scope = bucket.scope(config.scopeName);
+		this.collection = scope.collection(config.collectionName);
+	}
+
+	@Override
+	public void add(List<Document> documents) {
+		for (Document document : documents) {
+			if (Objects.isNull(document.getEmbedding()) || document.getEmbedding().isEmpty()) {
+				logger.debug("Calling embeddingModel for document id = " + document.getId());
+				document.setEmbedding(this.embeddingModel.embed(document));
+			}
+			collection.upsert(document.getId(), document);
+		}
+	}
+
+	@Override
+	public Optional<Boolean> delete(List<String> idList) {
+		for (String id : idList) {
+			collection.remove(id);
+		}
+		return Optional.empty();
+	}
+
+	@Override
+	public List<Document> similaritySearch(org.springframework.ai.vectorstore.SearchRequest springAiRequest) {
+		List<Double> queryEmbedding = this.embeddingModel.embed(springAiRequest.getQuery());
+
+		Float[] data = queryEmbedding.stream().map(Double::floatValue).toArray(Float[]::new);
+		float[] embeddings = new float[data.length];
+
+		for (int x = 0; x < data.length; x++) {
+			embeddings[x] = data[x].floatValue();
+		}
+
+		int topK = springAiRequest.getTopK();
+
+		double similarityThreshold = springAiRequest.getSimilarityThreshold();
+		Filter.Expression fe = springAiRequest.getFilterExpression();
+
+		String nativeFilterExpression = (fe != null) ? " AND " + this.filterExpressionConverter.convertExpression(fe)
+				: "";
+		String statement = String.format(
+				"""
+						SELECT c.* FROM `%s` AS c
+						WHERE SEARCH(`c`, {"query": {"match_none": {}}, "knn": [{"field": "embedding", "k": %s, "vector": %s }   ]    }, {"index": "%s.%s.%s"}   )
+						%s
+						""",
+				this.config.collectionName, topK, Arrays.toString(embeddings), this.config.bucketName,
+				this.config.scopeName, this.config.vectorIndexName, nativeFilterExpression);
+
+		QueryResult result = scope.query(statement, QueryOptions.queryOptions());
+
+		return result.rowsAs(Document.class);
+	}
+
+	@Override
+	public void afterPropertiesSet() {
+
+		if (!this.initializeSchema) {
+			return;
+		}
+
+		try {
+			initCluster();
+		}
+		catch (InterruptedException e) {
+			throw new RuntimeException(e);
+		}
+	}
+
+	public void initCluster() throws InterruptedException {
+		// init scope, collection, indexes
+		BucketSettings bs = cluster.buckets().getAllBuckets().get(this.config.bucketName);
+		if (bs == null) {
+			cluster.buckets().createBucket(BucketSettings.create(this.config.bucketName));
+		}
+		String fullyQualifieScopeName = String.format("`%s`.`%s`", this.config.bucketName, this.config.scopeName);
+		String fullyQualifiedCollectionName = String.format("`%s`.`%s`.`%s`", this.config.bucketName,
+				this.config.scopeName, this.config.collectionName);
+		Bucket b = cluster.bucket(this.config.bucketName);
+		Scope s = b.defaultScope();
+		s.query("CREATE SCOPE " + fullyQualifieScopeName + " IF NOT EXISTS;");
+		Thread.sleep(100);
```

Comment:
Why do we need to `sleep` here? Can we avoid that and use a more deterministic mechanism? There might be a scenario in which the code we want to execute fully may not be ready with the sleep.